api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +78,295 @@ export default async function bootChallenge(app, done) {
+      : null;
+    if (alreadyCompleted) {
+      $set[`completedChallenges.${oldIndex}`] = {
+        ...chal,
+        completedDate: oldChallenge.completedDate
+      };
+    } else {
+      $push.progressTimestamps = [...$push.progressTimestamps, Date.now()];
+      $push.completedChallenges = [...$push.completedChallenges, chal];
+    }
+  }
+
+  const updateData = {};
+  if ($push.completedChallenges.length) {
+    updateData.$push = $push;
+  }
+  if (!isEmpty($set)) {
+    updateData.$set = $set;
+  }
+
+  const updatePromise = new Promise((resolve, reject) =>
+    user.updateAttributes(updateData, err => {
+      if (err) {
+        return reject(err);
+      }
+      return resolve();
+    })
+  );
+  return Observable.fromPromise(updatePromise).doOnNext(() => {
+    return res.json({
+      completedChallenges
+    });
+  });
+}
+
+/**
+ * Handles submissions for all certification projects.
+ * @param {Request} req
+ * @param {Response} res
+ * @param {Next} _next
+ * @returns Good question??
+ */
+function projectCompleted(req, res, next) {
+  const { user, body } = req;
+
+  // Ensure `body` is an object
+  if (!body || typeof body !== 'object') {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid request format. Expected `body` to be an object.'
+    });
+  }
+
+  // `body` may exist as the following structures:
+  /*
+  # RWD Projects and JS Algorithm Projects
+  { 
+    challengeType: number,
+    files: [
+      {
+        contents: string,
+        ext: string,
+        history: string,
+        key: string,
+        name: string
+      }
+    ]
+    id: string,
+  }
+  # Front End Libraries and Data Visualisation and APIs
+  {
+    challengeType: number,
+    id: string,
+    solution: string,
+  }
+  # Back End Development / Quality Assurance / Scientific Computing with Python / Data Analysis with Python / Information Security / Machine Learning with Python
+  {
+    challengeType: number,
+    id: string,
+    solution: string,
+    githubLink?: string,
+  }
+  # Relational Database
+  {
+    challengeType: number,
+    id: string,
+    solution: string
+  }
+  # Take Home Projects
+  {
+    id: string,
+    solution: string,
+  }
+  */
+
+  const { id } = body;
+
+  // Find the structure, based on the `id` of the project
+  const isProjectValid = validateProject(id, body);
+
+  if (!isProjectValid) {
+    return res.status(403).json({
+      type: 'error',
+      message: 'Invalid project submission.'
+    });
+  }
+
+  // TODO: Handle all the different types of project
+  // TODO: Handle resubmission of projects (no new timestamp)
+  const updatedProperties = {};
+}
+
+/**
+ * From Naomi:
+ * TODO: Shaun, we'll need to update this to not use the challenge types from the client.
+ * We want to decouple the API and client, and this recouples it. Maybe a question for Oliver, but
+ * can we move the challengeTypes stuff into the root level utils?? Otherwise, we'll need to hard code here
+ * IMHO.
```

Comment:
@ojeytonwilliams Shall I hard-code this now? Or, move the challengeTypes to `/utils`?

Reply:
Moving it to `/utils` makes the most sense to me. We're already compiling any ts files in there, so it should be fine.